quiche/src/pmtud.rs

```diff
@@ -1,92 +1,193 @@
     }
 }
 
 impl std::fmt::Debug for Pmtud {
     fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
-        write!(f, "current={:?} ", self.cur_size)?;
-        write!(f, "probe_size={:?} ", self.probe)?;
-        write!(f, "continue_probing={:?} ", self.next_size)?;
-        write!(f, "enable={:?} ", self.enable)?;
+        write!(f, "current_mtu={:?} ", self.current_mtu)?;
+        write!(f, "probe_size={:?} ", self.probe_size)?;
+        write!(f, "should_probe={:?} ", self.should_probe)?;
+        write!(f, "enabled={:?} ", self.enabled)?;
         Ok(())
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn pmtud_initial_state() {
+        let mut pmtud = Pmtud::new(1200);
+
+        assert_eq!(pmtud.get_current_mtu(), 1200);
+        assert_eq!(pmtud.get_probe_size(), 0);
+        assert!(!pmtud.get_should_probe());
+        assert!(!pmtud.is_enabled());
+    }
+
+    #[test]
+    fn pmtud_enable_disable() {
+        let mut pmtud = Pmtud::new(1200);
+
+        pmtud.enable(true);
+        assert!(pmtud.is_enabled());
+
+        pmtud.enable(false);
+        assert!(!pmtud.is_enabled());
+    }
+
+    #[test]
+    fn pmtud_probe_flag_management() {
+        let mut pmtud = Pmtud::new(1200);
+
+        // Initially should not probe
+        assert!(!pmtud.get_should_probe());
+
+        // Enable probing
+        pmtud.set_should_probe(true);
+        assert!(pmtud.get_should_probe());
+
+        // Disable probing
+        pmtud.set_should_probe(false);
+        assert!(!pmtud.get_should_probe());
+    }
+
+    #[test]
+    fn pmtud_binary_search_algorithm() {
+        let mut pmtud = Pmtud::new(1200);
+
+        // Set initial probe size to 1500
+        pmtud.set_probe_size(1500);
+        assert_eq!(pmtud.get_probe_size(), 1500);
+
+        // Simulate probe loss - should update to midpoint
+        pmtud.update_probe_size();
+        // Expected: 1200 + ((1500 - 1200) / 2) = 1200 + 150 = 1350
+        assert_eq!(pmtud.get_probe_size(), 1350);
+
+        // Another probe loss
+        pmtud.update_probe_size();
+        // Expected: 1200 + ((1350 - 1200) / 2) = 1200 + 75 = 1275
+        assert_eq!(pmtud.get_probe_size(), 1275);
+    }
+
+    #[test]
+    fn pmtud_probe_lost_behavior() {
+        let mut pmtud = Pmtud::new(1200);
+        pmtud.set_probe_size(1500);
+        pmtud.set_should_probe(false);
+
+        // Simulate probe loss
+        pmtud.pmtu_probe_lost();
+
+        // Should re-enable probing and adjust size
+        assert!(pmtud.get_should_probe());
+        assert_eq!(pmtud.get_probe_size(), 1350); // binary search result
+    }
+
+    #[test]
+    fn pmtud_successful_probe() {
+        let mut pmtud = Pmtud::new(1200);
+        pmtud.set_probe_size(1400);
+
+        // Simulate successful probe
+        pmtud.set_current_mtu(1400);
+
+        assert_eq!(pmtud.get_current_mtu(), 1400);
+    }
+
+    #[test]
+    fn pmtud_binary_search_convergence() {
```

Comment:
It would be good to write a test that finds an MTU of arbitrary size.

For example: initial 1200, probe 1500, actual MTU 1300.
Progression:
probe 1500 -> fail
probe 1350 -> fail
probe 1275 -> success
probe 1312 -> fail
probe 1293 -> success
probe 1302 -> fail
probe 1297 -> success
probe 1299 -> success
probe 1300 -> success
probe 1301 -> fail

Reply:
I don't think I follow.  Is the intent in the current implementation that we send out several size probes at the same time and set the MTU to the highest successful one?

I think we're trying to improve how the code handles a single probe size, no binary searching.

Reply:
As far as I understand, the current algorithm in quiche only sends a single probe at a time, starting at some maximum value (`max_send_udp_payload_size`). As soon as a probe is successful, PMTUD terminates. If a probe is lost, we reduce probe size and try again. We never increase probe size after success